src/ui/castleinternaljoystickslinux.pas

```diff
@@ -145,54 +147,75 @@ procedure TLinuxJoysticksBackend.Poll(const List: TJoystickList;
                 begin
-                  Joystick.State.BtnUp[ event.number ] := True;
-                  Joystick.State.BtnPress   [ event.number ] := False;
-                  if Assigned(EventContainer.OnButtonUp) then EventContainer.OnButtonUp(Joystick, event.number);
-                  Joystick.State.BtnCanPress[ event.number ] := True;
-                end;
-
-                Joystick.State.BtnDown[ event.number ] := False;
-              end;
-            1:
-              begin
-                Joystick.State.BtnDown[ event.number ] := True;
-                if Assigned(EventContainer.OnButtonDown) then EventContainer.OnButtonDown(Joystick, event.number);
-                Joystick.State.BtnUp  [ event.number ] := False;
-                if Joystick.State.BtnCanPress[ event.number ] then
+                  if Joystick.State.BtnDown[ event.number ] then
                   begin
-                    Joystick.State.BtnPress   [ event.number ] := True;
-                    if Assigned(EventContainer.OnButtonPress) then EventContainer.OnButtonPress(Joystick, event.number);
-                    Joystick.State.BtnCanPress[ event.number ] := False;
+                    Joystick.State.BtnUp[ event.number ] := True;
+                    Joystick.State.BtnPress   [ event.number ] := False;
+                    if Assigned(EventContainer.OnButtonUp) then EventContainer.OnButtonUp(Joystick, event.number);
+                    Joystick.State.BtnCanPress[ event.number ] := True;
                   end;
-              end;
-          end;
+
+                  Joystick.State.BtnDown[ event.number ] := False;
+                end;
+              1:
+                begin
+                  Joystick.State.BtnDown[ event.number ] := True;
+                  if Assigned(EventContainer.OnButtonDown) then EventContainer.OnButtonDown(Joystick, event.number);
+                  Joystick.State.BtnUp  [ event.number ] := False;
+                  if Joystick.State.BtnCanPress[ event.number ] then
+                    begin
+                      Joystick.State.BtnPress   [ event.number ] := True;
+                      if Assigned(EventContainer.OnButtonPress) then EventContainer.OnButtonPress(Joystick, event.number);
+                      Joystick.State.BtnCanPress[ event.number ] := False;
+                    end;
+                end;
+            end;
+        end;
+        BytesRead := FpRead( BackendInfo.Device, event, 8 );
+      until BytesRead <> 8
+    else
+      if fpgeterrno <> EAGAIN then
+      begin
+        WritelnLog('Joystick error, possibly it was unplugged. Trying to run Joysticks.Initialize.');
+        JoystickHadBeenDisconnected := true;
+        { Note that only last unplugged joystick will be sent to the Unplugged event
+          if multiple joysticks have been unplugged simultaneously
+          which may happen e.g. in case an USB-Hub with several joysticks had been disconnected }
+        JoystickDisconnected := Joystick;
```

Comment:
Since it's not reliable (we could have multiple disconnections simultaneously, as you write) I would say to not try to remember/pass the `JoystickDisconnected`. I.e. let's not implement something we don't immediately need, when  we know the initial implementation would not be 100% reliable :) This way also code is a bit simpler.

So `Joysticks.OnDisconnect` would be parameter-less.

In the future we can make it better, e.g. collecting list of disconnected joysticks, if it will be ever needed.